### ISSUE TITLE
[feature] Support for Mac Catalyst

### ISSUE THREAD

**Issue by paulo-coutinho** [conan-io/conan#6259]
Hi,

Can team add support for Mac Catalyst? I read that arch for Mac Catalyst is "x86_64h".

Some docs:
- https://onesignal.com/blog/mac-catalyst-fat-framework/
- https://forums.developer.apple.com/thread/120542
- https://github.com/leetal/ios-cmake/issues/53

Im adding support for it into my toolchain: https://github.com/ezored/conan-darwin-toolchain

But when i try build with different arch from conan i get error:
```
conans.errors.ConanException: Invalid setting 'x86_64h' is not a valid 'settings.arch' value.
Possible values are ['armv4', 'armv4i', 'armv5el', 'armv5hf', 'armv6', 'armv7', 'armv7hf', 'armv7k', 'armv7s', 'armv8', 'armv8.3', 'armv8_32', 'asm.js', 'avr', 'mips', 'mips64', 'ppc32', 'ppc32be', 'ppc64', 'ppc64le', 's390', 's390x', 'sh4le', 'sparc', 'sparcv9', 'wasm', 'x86', 'x86_64']
Read "http://docs.conan.io/en/latest/faq/troubleshooting.html#error-invalid-setting"
```

With test:
```
builder = ConanMultiPackager()
builder.add(settings={"os": "iOS", "os.version": "13.0", "arch": "x86_64h"})
builder.run()
```

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

Thanks.

**Comment by jgsogo** [comment]
Hi, @prsolucoes 

Have you tried adding it yourself to your `settings.yml`? Have you had a look to the build helpers and tools in Conan and see how many of them need to be changed to consider that new architecture properly?

Thanks!

**Comment by prsolucoes** [comment]
Hi,

Yes, i already tested with my toolchain (https://github.com/ezored/conan-darwin-toolchain). You can see build here: https://github.com/ezored/conan-darwin-toolchain/blob/testing/1.1.0/build.py#L10

It is a work around because conan don't accept x86_64h.

Mac Catalyst is recent, only release with iOS 13 and latest xcode 11, so i don't know the public. It you don't understand that need change conan, no problem, im only pointing that have a new arch.

Thanks.

**Comment by prsolucoes** [comment]
Hi,

I have added x86_64h and it works well. See generated files:

```
/Users/paulo/Developer/workspaces/cpp/ezored/dist/ios_framework/Debug/Ezored.framework/Ezored: Mach-O universal binary with 6 architectures: [x86_64:Mach-O 64-bit dynamically linked shared library x86_64] [x86_64h] [arm_v7] [arm_v7s] [arm64] [arm64e]
/Users/paulo/Developer/workspaces/cpp/ezored/dist/ios_framework/Debug/Ezored.framework/Ezored (for architecture x86_64):	Mach-O 64-bit dynamically linked shared library x86_64
/Users/paulo/Developer/workspaces/cpp/ezored/dist/ios_framework/Debug/Ezored.framework/Ezored (for architecture x86_64h):	Mach-O 64-bit dynamically linked shared library x86_64h
/Users/paulo/Developer/workspaces/cpp/ezored/dist/ios_framework/Debug/Ezored.framework/Ezored (for architecture armv7):	Mach-O dynamically linked shared library arm_v7
/Users/paulo/Developer/workspaces/cpp/ezored/dist/ios_framework/Debug/Ezored.framework/Ezored (for architecture armv7s):	Mach-O dynamically linked shared library arm_v7s
/Users/paulo/Developer/workspaces/cpp/ezored/dist/ios_framework/Debug/Ezored.framework/Ezored (for architecture arm64):	Mach-O 64-bit dynamically linked shared library arm64
/Users/paulo/Developer/workspaces/cpp/ezored/dist/ios_framework/Debug/Ezored.framework/Ezored (for architecture arm64e):	Mach-O 64-bit dynamically linked shared library arm64e
```

I use into my project ezored (https://ezored.com/). Will be nice if can be included.

**Comment by jgsogo** [comment]
My main concern is about people using tools like [`to_apple_arch`](https://github.com/conan-io/conan/blob/337cd7214f46dcc8abca99acad652e448401c279/conans/client/tools/apple.py#L16) or other where we are mapping between Conan settings and the compiler/os language and flags. These usages can appear in tools, build helpers and generators. We need to have a look at all these places in order to add and support this new architecture in the settings provided by Conan.

But, yes, if it is used we should consider it 👍

**Comment by prsolucoes** [comment]
Hi,

I have changed this file too. But the changes are only two to make it work, these are the steps that i used locally:

1. Change settings to include `"x86_64h"`
2. Add line `'x86_64h': 'x86_64h',` on file: https://github.com/conan-io/conan/blob/develop/conans/client/tools/apple.py#L16

Nothing more.

**Comment by prsolucoes** [comment]
I think that it is not necessary because on xcode sample project that i created it don't use x86_64h:

COMPILE

```
CompileC /Users/paulo/Library/Developer/Xcode/DerivedData/LibCatalyst-dlmxxsedvbjqnxdqckncmnxmwcuk/Build/Intermediates.noindex/LibCatalyst.build/Debug-maccatalyst/LibCatalyst.build/Objects-normal/x86_64/LibCatalyst_vers.o /Users/paulo/Library/Developer/Xcode/DerivedData/LibCatalyst-dlmxxsedvbjqnxdqckncmnxmwcuk/Build/Intermediates.noindex/LibCatalyst.build/Debug-maccatalyst/LibCatalyst.build/DerivedSources/LibCatalyst_vers.c normal x86_64 c com.apple.compilers.llvm.clang.1_0.compiler (in target 'LibCatalyst' from project 'LibCatalyst')
    cd /Users/paulo/Developer/workspaces/xcode/LibCatalyst
    export LANG=en_US.US-ASCII
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -x c -target x86_64-apple-ios13.2-macabi -fmessage-length=0 -fdiagnostics-show-note-include-stack -fmacro-backtrace-limit=0 -std=gnu11 -fmodules -gmodules -fmodules-cache-path=/Users/paulo/Library/Developer/Xcode/DerivedData/ModuleCache.noindex -fmodules-prune-interval=86400 -fmodules-prune-after=345600 -fbuild-session-file=/Users/paulo/Library/Developer/Xcode/DerivedData/ModuleCache.noindex/Session.modulevalidation -fmodules-validate-once-per-build-session -Wnon-modular-include-in-framework-module -Werror=non-modular-include-in-framework-module -fmodule-name=LibCatalyst -Wno-trigraphs -fpascal-strings -O0 -fno-common -Wno-missing-field-initializers -Wno-missing-prototypes -Werror=return-type -Wdocumentation -Wunreachable-code -Werror=deprecated-objc-isa-usage -Werror=objc-root-class -Wno-missing-braces -Wparentheses -Wswitch -Wunused-function -Wno-unused-label -Wno-unused-parameter -Wunused-variable -Wunused-value -Wempty-body -Wuninitialized -Wconditional-uninitialized -Wno-unknown-pragmas -Wno-shadow -Wno-four-char-constants -Wno-conversion -Wconstant-conversion -Wint-conversion -Wbool-conversion -Wenum-conversion -Wno-float-conversion -Wnon-literal-null-conversion -Wobjc-literal-conversion -Wshorten-64-to-32 -Wpointer-sign -Wno-newline-eof -DDEBUG=1 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -fasm-blocks -fstrict-aliasing -Wdeprecated-declarations -g -Wno-sign-conversion -Winfinite-recursion -Wcomma -Wblock-capture-autoreleasing -Wstrict-prototypes -Wno-semicolon-before-method-body -Wunguarded-availability -index-store-path /Users/paulo/Library/Developer/Xcode/DerivedData/LibCatalyst-dlmxxsedvbjqnxdqckncmnxmwcuk/Index/DataStore -iquote /Users/paulo/Library/Developer/Xcode/DerivedData/LibCatalyst-dlmxxsedvbjqnxdqckncmnxmwcuk/Build/Intermediates.noindex/LibCatalyst.build/Debug-maccatalyst/LibCatalyst.build/LibCatalyst-generated-files.hmap -I/Users/paulo/Library/Developer/Xcode/DerivedData/LibCatalyst-dlmxxsedvbjqnxdqckncmnxmwcuk/Build/Intermediates.noindex/LibCatalyst.build/Debug-maccatalyst/LibCatalyst.build/LibCatalyst-own-target-headers.hmap -I/Users/paulo/Library/Developer/Xcode/DerivedData/LibCatalyst-dlmxxsedvbjqnxdqckncmnxmwcuk/Build/Intermediates.noindex/LibCatalyst.build/Debug-maccatalyst/LibCatalyst.build/LibCatalyst-all-non-framework-target-headers.hmap -ivfsoverlay /Users/paulo/Library/Developer/Xcode/DerivedData/LibCatalyst-dlmxxsedvbjqnxdqckncmnxmwcuk/Build/Intermediates.noindex/LibCatalyst.build/Debug-maccatalyst/LibCatalyst.build/all-product-headers.yaml -iquote /Users/paulo/Library/Developer/Xcode/DerivedData/LibCatalyst-dlmxxsedvbjqnxdqckncmnxmwcuk/Build/Intermediates.noindex/LibCatalyst.build/Debug-maccatalyst/LibCatalyst.build/LibCatalyst-project-headers.hmap -I/Users/paulo/Library/Developer/Xcode/DerivedData/LibCatalyst-dlmxxsedvbjqnxdqckncmnxmwcuk/Build/Products/Debug-maccatalyst/include -isystem /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/iOSSupport/usr/include -I/Users/paulo/Library/Developer/Xcode/DerivedData/LibCatalyst-dlmxxsedvbjqnxdqckncmnxmwcuk/Build/Intermediates.noindex/LibCatalyst.build/Debug-maccatalyst/LibCatalyst.build/DerivedSources-normal/x86_64 -I/Users/paulo/Library/Developer/Xcode/DerivedData/LibCatalyst-dlmxxsedvbjqnxdqckncmnxmwcuk/Build/Intermediates.noindex/LibCatalyst.build/Debug-maccatalyst/LibCatalyst.build/DerivedSources/x86_64 -I/Users/paulo/Library/Developer/Xcode/DerivedData/LibCatalyst-dlmxxsedvbjqnxdqckncmnxmwcuk/Build/Intermediates.noindex/LibCatalyst.build/Debug-maccatalyst/LibCatalyst.build/DerivedSources -F/Users/paulo/Library/Developer/Xcode/DerivedData/LibCatalyst-dlmxxsedvbjqnxdqckncmnxmwcuk/Build/Products/Debug-maccatalyst -iframework /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/iOSSupport/System/Library/Frameworks -MMD -MT dependencies -MF /Users/paulo/Library/Developer/Xcode/DerivedData/LibCatalyst-dlmxxsedvbjqnxdqckncmnxmwcuk/Build/Intermediates.noindex/LibCatalyst.build/Debug-maccatalyst/LibCatalyst.build/Objects-normal/x86_64/LibCatalyst_vers.d --serialize-diagnostics /Users/paulo/Library/Developer/Xcode/DerivedData/LibCatalyst-dlmxxsedvbjqnxdqckncmnxmwcuk/Build/Intermediates.noindex/LibCatalyst.build/Debug-maccatalyst/LibCatalyst.build/Objects-normal/x86_64/LibCatalyst_vers.dia -c /Users/paulo/Library/Developer/Xcode/DerivedData/LibCatalyst-dlmxxsedvbjqnxdqckncmnxmwcuk/Build/Intermediates.noindex/LibCatalyst.build/Debug-maccatalyst/LibCatalyst.build/DerivedSources/LibCatalyst_vers.c -o /Users/paulo/Library/Developer/Xcode/DerivedData/LibCatalyst-dlmxxsedvbjqnxdqckncmnxmwcuk/Build/Intermediates.noindex/LibCatalyst.build/Debug-maccatalyst/LibCatalyst.build/Objects-normal/x86_64/LibCatalyst_vers.o
```

LD

```
Ld /Users/paulo/Library/Developer/Xcode/DerivedData/LibCatalyst-dlmxxsedvbjqnxdqckncmnxmwcuk/Build/Products/Debug-maccatalyst/LibCatalyst.framework/Versions/A/LibCatalyst normal x86_64 (in target 'LibCatalyst' from project 'LibCatalyst')
    cd /Users/paulo/Developer/workspaces/xcode/LibCatalyst
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -target x86_64-apple-ios13.2-macabi -dynamiclib -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -L/Users/paulo/Library/Developer/Xcode/DerivedData/LibCatalyst-dlmxxsedvbjqnxdqckncmnxmwcuk/Build/Products/Debug-maccatalyst -L/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/iOSSupport/usr/lib -L/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/maccatalyst -L/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/iOSSupport/usr/lib -L/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/maccatalyst -F/Users/paulo/Library/Developer/Xcode/DerivedData/LibCatalyst-dlmxxsedvbjqnxdqckncmnxmwcuk/Build/Products/Debug-maccatalyst -iframework /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/iOSSupport/System/Library/Frameworks -iframework /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/iOSSupport/System/Library/Frameworks -filelist /Users/paulo/Library/Developer/Xcode/DerivedData/LibCatalyst-dlmxxsedvbjqnxdqckncmnxmwcuk/Build/Intermediates.noindex/LibCatalyst.build/Debug-maccatalyst/LibCatalyst.build/Objects-normal/x86_64/LibCatalyst.LinkFileList -install_name @rpath/LibCatalyst.framework/Versions/A/LibCatalyst -Xlinker -rpath -Xlinker @executable_path/Frameworks -Xlinker -rpath -Xlinker @loader_path/Frameworks -Xlinker -object_path_lto -Xlinker /Users/paulo/Library/Developer/Xcode/DerivedData/LibCatalyst-dlmxxsedvbjqnxdqckncmnxmwcuk/Build/Intermediates.noindex/LibCatalyst.build/Debug-maccatalyst/LibCatalyst.build/Objects-normal/x86_64/LibCatalyst_lto.o -Xlinker -export_dynamic -Xlinker -no_deduplicate -compatibility_version 1 -current_version 1 -Xlinker -dependency_info -Xlinker /Users/paulo/Library/Developer/Xcode/DerivedData/LibCatalyst-dlmxxsedvbjqnxdqckncmnxmwcuk/Build/Intermediates.noindex/LibCatalyst.build/Debug-maccatalyst/LibCatalyst.build/Objects-normal/x86_64/LibCatalyst_dependency_info.dat -o /Users/paulo/Library/Developer/Xcode/DerivedData/LibCatalyst-dlmxxsedvbjqnxdqckncmnxmwcuk/Build/Products/Debug-maccatalyst/LibCatalyst.framework/Versions/A/LibCatalyst
```

So, i think what i need in this thread is unnecessary.

**Comment by prsolucoes** [comment]
Hi,

Here is one company that is using `x86_64h` too:
https://github.com/OneSignal/OneSignal-iOS-SDK/blob/master/iOS_SDK/OneSignalSDK/build_fat_framework.sh

**Comment by prsolucoes** [comment]
Hi,

I was able to compile without x86_64h. It is here working on mobile, iPad, simulator and desktop:
https://github.com/ezored/ezored/issues/20

Im only waiting now for a patch to sqlite3 that i report on the same link. With that patch it will compile OK.

I made a video:
https://www.youtube.com/watch?v=ro2WVX_wHuQ

Thanks.

**Comment by prsolucoes** [comment]
Hi,

Conclusion is that we need compile for x86_64 but use this flags:
```
-target x86_64-apple-ios-macabi -miphoneos-version-min=13.0
```

Target OS need be Macos instead of iOS. Example:
https://github.com/ezored/conan-darwin-toolchain/blob/stable/1.1.0/build.py#L11-L21

All dependencies need be compiled with flags:
```
-target x86_64-apple-ios-macabi -miphoneos-version-min=13.0
```

I don't know if need some conan changes in settings and subsettings.

**Comment by SSE4** [comment]
probably, we need new os setting or sub-setting here. Mac catalyst libraries aren't compatible with iOS or MacOS native libraries, they have to be compiled with special flags:
```
-target x86_64-apple-ios-macabi -miphoneos-version-min=13.0
```

**Comment by kaptenhonek** [comment]
I would love to see support for this as well

**Comment by dmn-star** [comment]
We plan release our Catalyst app for the beginning of the next year. Is there a timeline for this feature?

**Comment by memsharded** [comment]
Hi @dmn-star 

So far this feature didn't get enough priority. It seems that @prsolucoes did a good job adding to their settings and customizing their toolchains, and there weren't a strong evidence of many users wanting this to increase the priority. If everybody here things it is worth it, I wouldn't oppose to try to help from our side, because it seems we don't have the time resources and experience. I would also request here to try to implement it in the new toolchains if possible, to align with our current efforts and future evolution of the integration with the build systems.

**Comment by dmn-star** [comment]
Hi @prsolucoes 

I see, you have removed support for catalyst from your toolchain. Was there a particular reason for this?

**Comment by paulocoutinhox** [comment]
Hi,

Nobody use it and now all iOS apps work on arm64 desktop.

Thanks.

**Comment by dmn-star** [comment]
That's a good point. Unfortunately I have to port our iOS app to Mac catalyst.

**Comment by memsharded** [comment]
Merged #8264, will be released soon in 1.33
